src/main/java/me/zeroeightsix/fiber/api/schema/type/derived/ConfigTypes.java

```diff
@@ -0,0 +1,145 @@
+
+    /* Enum-derived types */
+
+    public static <E extends Enum<E>> EnumConfigType<E> makeEnum(Class<E> enumType) {
+        Set<String> validValues = Arrays.stream(enumType.getEnumConstants()).map(Enum::name).collect(Collectors.toSet());
+        return new EnumConfigType<>(new EnumSerializableType(validValues), enumType, e -> Enum.valueOf(enumType, e), Enum::name);
+    }
+
+    /* List-derived types */
+
+    public static <E0, E, U extends ConfigType<E, E0, ?>> ListConfigType<List<E>, E0> makeList(U elementType) {
+        return new ListConfigType<>(
+                new ListSerializableType<>(elementType.getSerializedType()), List.class,
+                l -> {
+                    List<E> ret = new ArrayList<>();
+                    for (E0 e0 : l) {
+                        ret.add(elementType.toRuntimeType(e0));
+                    }
+                    return Collections.unmodifiableList(ret);
+                },
+                l -> {
+                    List<E0> ret = new ArrayList<>();
+                    for (E e : l) {
+                        ret.add(elementType.toPlatformType(e));
+                    }
+                    return Collections.unmodifiableList(ret);
+                }
+        );
+    }
+
+    public static <E0, E, U extends ConfigType<E, E0, ?>> ListConfigType<Set<E>, E0> makeSet(U elementType) {
+        return new ListConfigType<>(
+                new ListSerializableType<>(elementType.getSerializedType(), 0, Integer.MAX_VALUE, true),
+                Set.class,
+                l -> {
+                    Set<E> ret = new HashSet<>();
+                    for (E0 e0 : l) {
+                        ret.add(elementType.toRuntimeType(e0));
+                    }
+                    return Collections.unmodifiableSet(ret);
+                },
+                l -> {
+                    List<E0> ret = new ArrayList<>();
+                    for (E e : l) {
+                        ret.add(elementType.toPlatformType(e));
+                    }
+                    return Collections.unmodifiableList(ret);
+                }
+        );
+    }
+
+    public static <E0, E, U extends ConfigType<E, E0, ? extends SerializableType<E0>>> ListConfigType<E[], E0> makeArray(U elementType) {
+        @SuppressWarnings("unchecked") Class<E[]> arrType = (Class<E[]>) Array.newInstance(elementType.getRuntimeType(), 0).getClass();
+        return new ListConfigType<>(
+                new ListSerializableType<>(elementType.getSerializedType()),
+                arrType,
+                l -> {
+                    @SuppressWarnings("unchecked") E[] ret = (E[]) Array.newInstance(elementType.getRuntimeType(), l.size());
+                    for (int i = 0; i < l.size(); i++) {
+                        ret[i] = elementType.toRuntimeType(l.get(i));
+                    }
+                    return ret;
+                },
+                arr -> {
```

Comment:
There is an implicit cast here that may also fail when the runtyoe type represents an array of primitive.